website/docs/script/type-id.mdx

```diff
@@ -0,0 +1,237 @@
+#define SCRIPT_SIZE 32768
+
+int main() {
+  uint64_t len = 0;
+  int ret = ckb_load_cell(NULL, &len, 0, 1, CKB_SOURCE_GROUP_OUTPUT);
+  if (ret != CKB_INDEX_OUT_OF_BOUND) {
+    return -1;  /* 1 */
+  }
+
+  len = 0;
+  ret = ckb_load_cell(NULL, &len, 0, 0, CKB_SOURCE_GROUP_INPUT);
+  if (ret != CKB_INDEX_OUT_OF_BOUND) {
+    return 0;  /* 2 */
+  }
+
+  /* 3 */
+  unsigned char input[INPUT_SIZE];
+  uint64_t input_len = INPUT_SIZE;
+  ret = ckb_load_input(input, &input_len, 0, 0, CKB_SOURCE_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (input_len > INPUT_SIZE) {
+    return -100;
+  }
+
+  unsigned char Script[SCRIPT_SIZE];
+  len = SCRIPT_SIZE;
+  ret = ckb_load_script(Script, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (len > SCRIPT_SIZE) {
+    return -101;
+  }
+  mol_seg_t script_seg;
+  script_seg.ptr = (uint8_t *)Script;
+  script_seg.size = len;
+
+  if (MolReader_Script_verify(&script_seg, false) != MOL_OK) {
+    return -102;
+  }
+
+  mol_seg_t args_seg = MolReader_Script_get_args(&script_seg);
+  mol_seg_t args_bytes_seg = MolReader_Bytes_raw_bytes(&args_seg);
+
+  if ((input_len == args_bytes_seg.size) &&
+      (memcmp(args_bytes_seg.ptr, input, input_len) == 0)) {
+    /* 4 */
+    return 0;
+  }
+  return -103;
+}
+```
+
+Attackers will be prevented in several different directions:
+
+1. If an attacker tries to create a cell with the exactly same type Script, there will be 2 cases:
+   a. A valid transaction will have different OutPoint data in the first input from the type Script args;
+   b. If the user tries to duplicate type Script args as the first transaction input, CKB will signal a double-spent error;
+2. If the attacker tries to use a different type Script args, it will be a different type Script by definition.
+
+This way, we can ensure a cell will have unique type Script across all live cells in CKB. Considering each Script has an associated hash, we will have a cell in CKB with its unique hash, or unique ID.
+
+## Resolving Scripts in CKB Transaction
+
+Now let's look at how CKB resolves the Script to run before the `hash_type` change:
+
+- CKB extracts `code_hash` value from the Script structure to run.
+- It loops through all dep cells, computes the hash of each dep cell data. If any dep cell data hash matches the specified `code_hash`, CKB uses the data in the found dep cell as the Script to run.
+- If no dep cell has data hash matching specified `code_hash`, CKB results in a validation error.
+
+The upgradability problem, in fact lies in the way we test for dep cells. Right now we are testing against data hash, and when a Script is upgraded, a different hash will be generated, the matching would fail. That brings a question: can we test for dep cells using a different solution? Is there something that can stay unchanged when the Script is changed? Considering the actual Script to run lives in a cell, we can rephrase the question in a different way:
```

Comment:
```suggestion
The upgradability problem arises because we test dep Cells against data hashes. When a Script is upgraded, the hash changes, causing the match to fail. Therefore, we need a different approach to test dep Cells. 
```